sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs

```diff
@@ -4101,6 +4115,152 @@ private void SetNameFieldsIfNull()
             }
         }
         #endregion
+
+        #region GenerateSAS
+        /// <summary>
+        /// The <see cref="GetSasBuilder"/> returns a <see cref="BlobSasBuilder"/> that
+        /// sets the respective properties in the BlobSasBuilder from the client.
+        /// </summary>
+        /// <param name="permissions">
+        /// Specifies the list of permissions that can be set in the SasBuilder
+        /// See <see cref="BlobSasPermissions"/>.
+        /// </param>
+        /// <param name="expiresOn">
+        /// Specifies when to set the expires time in the sas builder
+        /// </param>
+        /// <returns>
+        /// A <see cref="BlobSasBuilder"/> on successfully deleting.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        public BlobSasBuilder GetSasBuilder(
+            BlobSasPermissions permissions,
+            DateTimeOffset expiresOn)
+        {
+            BlobSasBuilder sasBuilder = new BlobSasBuilder
+            {
+                Version = Version.ToString(),
+                BlobContainerName = BlobContainerName,
+                BlobName = Name,
+                ExpiresOn = expiresOn,
+                Resource = "b"
+            };
+            sasBuilder.SetPermissions(permissions);
+            return sasBuilder;
+        }
+
+        /// <summary>
+        /// The <see cref="GenerateSasUri"/> returns a Uri that
+        /// generates a Service SAS based on the Client properties and builder passed.
+        ///
+        /// For more information, see
+        /// <see href="https://docs.microsoft.com/en-us/rest/api/storageservices/constructing-a-service-sas">
+        /// Consturcting a Service SAS</see>
+        /// </summary>
+        /// <param name="builder">
+        /// Used to generate a Shared Access Signature (SAS)
+        /// </param>
+        /// <returns>
+        /// A <see cref="BlobSasBuilder"/> on successfully deleting.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        public Uri GenerateSasUri(
+            BlobSasBuilder builder)
```

Comment:
Do we need this API? I think blobClient->Builder->SAS sounds like natural flow but using blobClient to inject credentials into externally created builder sounds backwards. I'd rather remove that unless there's specific ask for such functionality (if there is why don't we make API on SAS builder to derive credential from client, i.e. reverse this flow).

Reply:
I think we decided to go this route because we didn't want to create a new class that was exactly like the `BlobSasBuilder`. but we still wanted to have a class that encompasses all the parameters for the sasBuilder.

There's also been an ask not to pass around the credential and the client, and keep it under the hood, since it already has taken it in at some point. https://github.com/Azure/azure-sdk-for-net/issues/11055

If the flow was like builder.generateSas(blobClient), we would have to make the credential public which I think the arch board has not been a fan of, and it might be best to keep it internal to the client.

Reply:
Couldn't BlobClient and BlobSasBuilder exchange credentials through internals? 
Something like builder.ForBlobClient(blobCLient).GenerateSAS() ?

Reply:
I'm not a fan of this proposal.  It complicates `BlobSasBuilder` to have a mode where you sometimes need credentials and sometimes don't.  Plus it's awkward to write `blob.GetSasBuilder(permissions, expiresOn).ForBlobClient(blob).GenerateSasUri()`.  And what does `ForBlobClient` do?  Just share credentials or set other properties and how do customers understand the relationship between that and `BlobClient.GetSasBuilder`?  I also think `BlobClient.GenerateSasUri` that takes a builder is easier to discover than a `ForBlobClient` method on the builder itself.

If anything maybe we should add an overload like `public Uri GenerateSasUri(BlobSasPermissions permissions, DateTimeOffset expiresOn) => GenerateSasUri(GetSasBuilder(permissions, expiresOn));` to make it as simple as humanly possible for customers to get a SAS URI to the blob/container they just created so they can pass it to whoever needs it next?  Then it's literally: `blob.Upload(myFile); Console.WriteLine(blob.GenerateSasUri(BlobSasPermissions.Read, DateTime.UtcNow.AddHours(3)))`.

Reply:
@tg-msft Yeah. The proposal was loosely defined suggestion how this could work if we forget about existing SAS builder shape. I agree that having two modes in SAS Builder is going to be confusing. I was rather thinking about how I would like to code against new API in ideal conditions. What I really dislike about API in this PR is that one has to start and terminate the flow at blobClient - which is kind of circular. Ideally I'd write `blobClient.GetSasBuilder(..params).GenerateUri()` or `new SasBuilder(blobClient).SetX().GenerateUri()`.

I think that `public Uri GenerateSasUri(BlobSasPermissions permissions, DateTimeOffset expiresOn)` overload is great idea - maybe this is the only API we should expose initially? I think we could also think about stretching this proposal a little bit. It seems that `BlobSasBuilder` is bunch of properties and code that generates final URL with some credential. Maybe we could extract `BlobSasOptions` that would encapsulate all these knobs and have `public Uri GenerateSasUri(BlobSasOptions)` ? Then BlobSasBuilder could inherit from `BlobSasOptions` to make sure it's kept consistent.